File: example-clients/typescript/poll_client.ts
/**
 * Consume a WorldSignal subscription by polling — no connection held open.
 * Env: WS_API_BASE, WS_API_KEY, WS_SUBSCRIPTION, WS_SINCE, WS_MAX, WS_INTERVAL.
 */
const BASE = process.env.WS_API_BASE ?? "http://localhost:4000";
const KEY = process.env.WS_API_KEY ?? "";
const SUB = process.env.WS_SUBSCRIPTION ?? "demo-stream";
const MAX = Number(process.env.WS_MAX ?? "0");
const INTERVAL = Number(process.env.WS_INTERVAL ?? "3") * 1000;

interface PollResponse { cursor: number; events: { seq: number; payload: { data?: { severity?: string; country?: string; title?: string } } }[] }

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

async function main(): Promise<void> {
  if (!KEY) { console.error("WS_API_KEY is required"); process.exit(2); }
  let cursor = Number(process.env.WS_SINCE ?? "0");
  let seen = 0;
  console.error(`[poll] ${BASE}/v1/stream/poll subscription=${SUB}`);
  for (;;) {
    const resp = await fetch(`${BASE}/v1/stream/poll?subscription=${SUB}&since=${cursor}`, {
      headers: { Authorization: `Bearer ${KEY}` },
    });
    if (!resp.ok) { console.error(`[poll] http ${resp.status}`); process.exit(1); }
    const body = (await resp.json()) as PollResponse;
    for (const ev of body.events) {
      const d = ev.payload.data ?? {};
      console.log(`[poll] ${(d.severity ?? "?").padEnd(8)} ${d.country ?? "--"}  ${d.title ?? ""}`);
      if (MAX && ++seen >= MAX) process.exit(0);
    }
    cursor = body.cursor ?? cursor;
    await sleep(INTERVAL);
  }
}

void main();

File: example-clients/python/poll_client.py
#!/usr/bin/env python3
"""Consume a WorldSignal subscription by polling — no connection held open.

Each request returns the events after ?since=<cursor> plus the next cursor; the
client persists the cursor and polls again. Good for serverless / cron consumers.

Env: WS_API_BASE, WS_API_KEY, WS_SUBSCRIPTION, WS_SINCE, WS_MAX (see sse_client.py),
     WS_INTERVAL seconds between polls (default 3).
"""
import json
import os
import sys
import time
import urllib.request

BASE = os.environ.get("WS_API_BASE", "http://localhost:4800")
KEY = os.environ.get("WS_API_KEY", "")
SUB = os.environ.get("WS_SUBSCRIPTION", "demo-stream")
MAX = int(os.environ.get("WS_MAX", "0"))
INTERVAL = float(os.environ.get("WS_INTERVAL", "3"))


def poll(cursor: int):
    url = f"{BASE}/v1/stream/poll?subscription={SUB}&since={cursor}"
    req = urllib.request.Request(url, headers={"Authorization": f"Bearer {KEY}"})
    with urllib.request.urlopen(req) as resp:
        return json.load(resp)


def main() -> int:
    if not KEY:
        print("WS_API_KEY is required", file=sys.stderr)
        return 2
    cursor = int(os.environ.get("WS_SINCE", "0"))
    seen = 0
    print(f"[poll] {BASE}/v1/stream/poll subscription={SUB}", file=sys.stderr)
    while True:
        body = poll(cursor)
        for ev in body.get("events", []):
            d = ev["payload"].get("data", {})
            print(f"[poll] {d.get('severity','?'):8} {d.get('country') or '--'}  {d.get('title','')}")
            seen += 1
            if MAX and seen >= MAX:
                return 0
        cursor = body.get("cursor", cursor)
        time.sleep(INTERVAL)


if __name__ == "__main__":
    raise SystemExit(main())
